kernel/kernel-defaults/src/test/scala/io/delta/kernel/defaults/DeltaTableWritesSuite.scala

```diff
@@ -1321,6 +1321,369 @@ class DeltaTableWritesSuite extends DeltaTableWriteSuiteBase with ParquetSuiteBa
+      val schema = new StructType()
+        .add("a", StringType.STRING, true)
+        .add("c", IntegerType.INTEGER, true)
+
+      createTxn(engine, tablePath, isNewTable = true, schema, partCols = Seq.empty,
+        tableProperties = Map(TableConfig.COLUMN_MAPPING_MODE.getKey -> "id",
+          TableConfig.ICEBERG_COMPAT_V2_ENABLED.getKey -> "true")).commit(engine, emptyIterable())
+
+      val currentSchema = table.getLatestSnapshot(engine).getSchema
+      val newSchema = new StructType()
+        .add("a", StringType.STRING, true, currentSchema.get("a").getMetadata)
+        .add("b",
+          new StructType()
+            .add("non_nullable_field", IntegerType.INTEGER, false,
+              FieldMetadata.builder().putLong(ColumnMapping.COLUMN_MAPPING_ID_KEY, 4)
+                .putString(ColumnMapping.COLUMN_MAPPING_PHYSICAL_NAME_KEY, "d").build())
+            .add("e", IntegerType.INTEGER, true,
+              FieldMetadata.builder().putLong(ColumnMapping.COLUMN_MAPPING_ID_KEY, 5)
+                .putString(ColumnMapping.COLUMN_MAPPING_PHYSICAL_NAME_KEY, "e").build()), true,
+          FieldMetadata.builder().putLong(ColumnMapping.COLUMN_MAPPING_ID_KEY, 3)
+            .putString(ColumnMapping.COLUMN_MAPPING_PHYSICAL_NAME_KEY, "b").build())
+        .add("c", IntegerType.INTEGER, true, currentSchema.get("c").getMetadata)
+
+      val e = intercept[IllegalArgumentException] {
+        table.createTransactionBuilder(engine, testEngineInfo, Operation.MANUAL_UPDATE)
+          .withSchema(engine, newSchema)
+          .build(engine).commit(engine, emptyIterable())
+      }
+      assert(e.getMessage.contains(
+        "Cannot add a non-nullable field non_nullable_field"))
+    }
+  }
+
+  test("Test cannot drop a partition column") {
+    withTempDirAndEngine { (tablePath, engine) =>
+      val table = Table.forPath(engine, tablePath)
+      val schema = new StructType()
+        .add("a", StringType.STRING, true)
+        .add("c", IntegerType.INTEGER, true)
+
+      createTxn(engine, tablePath, isNewTable = true, schema, partCols = Seq("c"),
+        tableProperties = Map(TableConfig.COLUMN_MAPPING_MODE.getKey -> "id",
+          TableConfig.ICEBERG_COMPAT_V2_ENABLED.getKey -> "true")).commit(engine, emptyIterable())
+
+      val currentSchema = table.getLatestSnapshot(engine).getSchema
+      val newSchema = new StructType()
+        .add("a", StringType.STRING, true, currentSchema.get("a").getMetadata)
+        .add("b",
+          new StructType()
+            .add("d", IntegerType.INTEGER, true,
+              FieldMetadata.builder().putLong(ColumnMapping.COLUMN_MAPPING_ID_KEY, 4)
+                .putString(ColumnMapping.COLUMN_MAPPING_PHYSICAL_NAME_KEY, "d").build())
+            .add("e", IntegerType.INTEGER, true,
+              FieldMetadata.builder().putLong(ColumnMapping.COLUMN_MAPPING_ID_KEY, 5)
+                .putString(ColumnMapping.COLUMN_MAPPING_PHYSICAL_NAME_KEY, "e").build()), true,
```

Comment:
Maybe put some of these field metadata building in some helper so it's easier to read